src/ResolveIteratorCapableTrait.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Traversable;
+use Iterator;
+use IteratorAggregate;
+use Exception as RootException;
+use OutOfRangeException;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+
+/**
+ * Capable of resolving an iterable to its innermost iterable.
+ *
+ * @since [*next-version*]
+ */
+trait ResolveIteratorCapableTrait
+{
+    /**
+     * Finds the deepest iterator that matches.
+     *
+     * Because the given traversable can be an {@see IteratorAggregate}, it will try to get its inner iterator.
+     * On each iterator, it will run the test function.
+     * It will keep going inwards until one of these conditions occurs:
+     *      - The iterator matches (test function returns `true`).
+     *          This can also be the supplied iterator. In this case the iterator will be returned.
+     *      - The limit or the maximal depth are reached.
+     *          In this case, if the iterator is not a match, an exception will be thrown.
+     *
+     * @since [*next-version*]
+     *
+     * @param Traversable $iterator The iterator to resolve.
+     * @param callable    $test     The test function which determines when the iterator is considered to be resolved.
+     *                              Default: Returns `true` on first found instance of {@see Iterator}.
+     * @param $limit int|float|string|Stringable The depth limit for resolution.
+     *
+     * @throws InvalidArgumentException If limit is not a valid integer representation.
+     * @throws OutOfRangeException      If infinite recursion is detected, or the iterator could not be resolved within the depth limit.
+     *
+     * @return Iterator The inner-most iterator, or whatever the test function allows.
+     */
+    protected function _resolveIterator(Traversable $iterator, $test = null, $limit = null)
+    {
+        $origIterator = $iterator;
+
+        // Default test function
+        if (!is_callable($test)) {
+            $test = function (Traversable $subject) {
+                return $subject instanceof Iterator;
+            };
+        }
+
+        if (is_null($limit)) {
+            $limit = 100;
+        } else {
+            $limit = $this->_normalizeInt($limit);
+        }
+
+        $i = 0;
+        while (!$test($iterator) && $i < $limit) {
+            /* @var $iterator IteratorAggregate */
+            $_it = $iterator->getIterator();
```

Comment:
Is `$iterator` guaranteed to be an `IteratorAggregate` here? Perhaps I missed something. To me it seems like it could be anything, and the only definitive thing it can be is `Traversable`.

Reply:
1. `Traversable` is not "anything".
2. It is one of two things, and if it's not an `Iterator` then it is an `IteratorAggregate`.

However, I realize now that the test function may not include the test for `Iterator`, and

Reply:
The param is type hinted as `Traversable`.

It is copied into a temporary variable `$origIterator`.

In a `while` loop, it is tested against the `test` callback.

I don't see anything that says it's not an `Iterator`, and I don't see anything that says it _has_ to be an `IteratorAggregate`. I'm not saying this is the case. But I don't see anything that indicates otherwise. I'm just concerned about a potential loose assumption.